packages/compartment-mapper/src/import-archive.js

```diff
@@ -69,6 +79,30 @@ const makeArchiveImportHookMaker = (archive, compartments, archiveLocation) => {
       }
       const moduleLocation = `${packageLocation}/${module.location}`;
       const moduleBytes = await archive.read(moduleLocation);
+
+      if (computeSha512 !== undefined) {
+        if (module.sha512 !== undefined) {
+          const sha512 = computeSha512(moduleBytes);
+          if (sha512 !== module.sha512) {
+            throw new Error(
+              `Module ${q(module.location)} of package ${q(
+                packageLocation,
+              )} in archive ${q(
+                archiveLocation,
+              )} failed a SHA-512 integrity check`,
+            );
+          }
+        } else {
+          throw new Error(
```

Comment:
If I'm reading this correctly, if you 1: provide `computeSha512` into the archive importer and 2: the compartment map did not specify a hash for the module, then the import will fail. I'm all for things that encourage hashes, including throwing random errors if you choose to omit them, but this smells of an inconsistency.

* Providing the expected top-level compartment-map hash is a great signal to say that you care about the integrity of the thing you're loading.
  * Providing that hash without also providing the means to compute hashes is a clear mistake.
  * If you don't care about integrity, you wouldn't provide an expected compartment-map hash, but you might still provide the `computeSha512` tool, just like you'd unconditionally provide the file-reading tool.
* An archive which cares about integrity will include module hashes.
  * .. but maybe we're providing for archives which don't care about integrity, and lack those module hashes?
* An archive whose compartment-map does not match the expected hash is a clear error.

So I see three questions:
* If you provide `computeSha512`, but the archive does not include module hashes, I think this switch might fail.. should the application author (who may or may not have provided an expected compartment-map hash) think "oh, this archive should care about integrity more", or should they think "I shouldn't provide `computeSha512`" ?
* What should an application which cares about integrity do when asked to load an archive which does not?
* Should we even support the idea of an archive which does not care about integrity?

Reply:
I think I like best:

1. If you provide `computeSha512`, regardless of whether you provide `expectedSha512`, every module in `compartment-map.json` that provides a hash will be checked. Hashes are not required.
2. If you provide `computeSha512` and `expectedSha512`, the integrity of the `compartment-map.json` hash will be checked.
3. If you provide `computeSha512` but do not provide `expectedSha512`, the hash of `compartment-map.json` will not be checked.
4. If you do not provide `computeSha512` and do provide `expectedSha512`, this is a usage error.
5. If you provide neither `expectedSha512` nor `computeSha512`, no hashes will be checked, regardless of whether the compartment map has any hashes.